File: Arrays/disc_intersection/code.java

//  Logic
/*
The for loop iterates through the positions on the line.

At iteration i, the "active" discs (call these the set Ai) are 
those discs who have started (that is, their left boundary is to the left of i) 
but not ended (that is, their right boundary is on or to the right of i).

The number startArr[i] describes the number of discs which are 
activated at position i (meaning their left boundary is on position i;
 call these the set Bi).

It's clear that each of these discs in both sets mutually intersect. 
We know we have already counted the intersections of Ai with itself. 
However, we do need to count the intersections of Ai with Bi: 
this is the number active * startArr[i]. We also need to count the intersections 
of Bi with itself: this is the number startArr[i] * (startArr[i] - 1) / 2 since n∗(n−1)/2 
is the number of ways to pick two things from a set of size n.

Finally, after adding these intersections, it updates the number of active discs.
 Indeed, the number of active discs is precisely the old number plus the ones
 which started at position i and minus the ones that ended at position i.*/


 // you can also use imports, for example:
// import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");
*/
class Solution {
    public int solution(int[] A) {
       // write your code in Java SE 8
       int N=A.length;
       if(N==0)
       return 0;
       int sum=0;
       int start[]=new int[N];
       int end[]=new int[N];
       // sort start and end boundaries of the discs into their own arrays
       for(int i=0;i<N;i++)
       {
           if(A[i]>i)
           start[0]++;
           else
           start[i-A[i]]++;
           
           if(((long)i+(long)A[i]) >= (long)N)
           end[N-1]++;
           else
           end[i+A[i]]++;
           
       }
       int active=0;
       for(int i=0;i<N;i++)
       {
           sum+=active*start[i]+(start[i]*(start[i]-1))/2;
            // over 10 000 000 return -1 as stated in the assignment
           if(sum>10000000)
           return -1;
           active+=start[i]-end[i];
       }
       return sum;
   }
}

File: Dynamic Programming/FloodDepth/code.js
// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(A) {
    // write your code in JavaScript (Node.js 8.9.4)
    let N = A.length;
    if (N === 1)
        return 0;
    let left = new Array(N).fill(0);
    left[0] = A[0];
    for (let i = 1; i < N; i++) {
        left[i] = Math.max(left[i - 1], A[i]);
    }
    let right = new Array(N).fill(0);
    right[N - 1] = A[N - 1];
    for (let i = N - 2; i >= 0; i--) {
        right[i] = Math.max(right[i + 1], A[i]);
    }
    let res = 0;
    for (let i = 1; i < N - 1; i++) {
        let depth = Math.min(left[i] - A[i], right[i] - A[i]);
        if (res < depth)
            res = depth;
    }
    return res;
}